test/ered_tls_SUITE.erl

```diff
@@ -0,0 +1,287 @@
+
+    ConnectedInit = [#{msg_type := connected} = msg(addr, {"127.0.0.1", Port})
+                     || Port <- [Port1, Port2]],
+
+    #{slot_map := SlotMap} = msg(msg_type, slot_map_updated, 1000),
+
+    IdMap =  maps:from_list(lists:flatmap(
+                              fun([_,_|Nodes]) ->
+                                      [{Port, Id} || [_Addr, Port, Id |_]<- Nodes]
+                              end, SlotMap)),
+
+    ConnectedRest = [#{msg_type := connected} = msg(addr, {"127.0.0.1", Port})
+                     || Port <- PortsRest],
+
+    ClusterIds = [Id || #{cluster_id := Id} <- ConnectedInit ++ ConnectedRest],
+    ClusterIds = [maps:get(Port, IdMap) || Port <- Ports],
+
+    ?MSG(#{msg_type := cluster_ok}),
+
+    %% Clear all old data
+    [{ok, _} = ered:command_client(Client, [<<"FLUSHDB">>]) || Client <- ered:get_clients(P)],
+
+    no_more_msgs(),
+    P.
+
+msg(Key, Val) ->
+    msg(Key, Val, 1000).
+
+msg(Key, Val, Time) ->
+    receive
+        M = #{Key := Val} -> M
+    after Time ->
+            error({timeout, {Key, Val}, erlang:process_info(self(), messages)})
+    end.
+
+no_more_msgs() ->
+    {messages,Msgs} = erlang:process_info(self(), messages),
+    case  Msgs of
+        [] ->
+            ok;
+        Msgs ->
+            error({unexpected,Msgs})
+    end.
+
+cmd_log(Cmd) ->
+    R = os:cmd(Cmd),
+    ct:pal("~s\n~s\n", [Cmd, R]),
+    R.
+
+
+%% Basic test of commands when using TLS.
+t_command(_) ->
+    R = start_cluster(),
+    lists:foreach(fun(N) ->
+                          {ok, <<"OK">>} = ered:command(R, [<<"SET">>, N, N], N)
+                  end,
+                  [integer_to_binary(N) || N <- lists:seq(1,100)]),
+    no_more_msgs().
+
+%% Setup ered using an expired client certificate in TLSv1.2.
+t_expired_cert_tls_1_2(_) ->
+    generate_expired_client_cert(),
+
+    ClientOpts = [{connection_opts, [{tls_options, ?TLS_OPTS ++ [{versions, ['tlsv1.2']}]}]}],
+
+    {ok, _R} = ered:start_link([{"127.0.0.1", 31001}],
+                               [{info_pid, [self()]}, {client_opts, ClientOpts}]),
+
+    ?MSG(#{msg_type := connect_error, addr := {"127.0.0.1", 31001},
+           reason := {tls_alert,
+                      {certificate_expired, _}}}),
+    ?MSG(#{msg_type := node_down_timeout, addr := {"127.0.0.1", 31001}}, 2500),
+    no_more_msgs().
+
+%% Setup ered using an expired client certificate in TLSv1.3.
+t_expired_cert_tls_1_3(_) ->
+    generate_expired_client_cert(),
+
+    ClientOpts = [{connection_opts, [{tls_options, ?TLS_OPTS ++ [{versions, ['tlsv1.3']}]}]}],
+
+    {ok, _R} = ered:start_link([{"127.0.0.1", 31001}],
+                               [{info_pid, [self()]}, {client_opts, ClientOpts}]),
+
+    ?MSG(#{msg_type := socket_closed, addr := {"127.0.0.1", 31001},
+           reason := {recv_exit,
+                      {tls_alert,
+                       {certificate_expired, _}}}}),
```

Comment:
We get this reliably for TLS 1.3? That's great!

Reply:
Seem to work fine so far.